src/components/paster.ts

```diff
@@ -0,0 +1,201 @@
+import * as vscode from 'vscode'
+import * as path from 'path'
+import * as fs from 'fs'
+import * as csv from 'csv-parser'
+
+import { Extension } from '../main'
+
+export class Paster {
+    extension: Extension
+
+    constructor(extension: Extension) {
+        this.extension = extension
+    }
+
+    public async paste() {
+        this.extension.logger.addLogMessage('Performing formatted paste')
+
+        // get current edit file path
+        const editor = vscode.window.activeTextEditor
+        if (!editor) {
+            return
+        }
+
+        const fileUri = editor.document.uri
+        if (!fileUri) {
+            return
+        }
+        if (fileUri.scheme === 'untitled') {
+            vscode.window.showInformationMessage('Before paste image, you need to save current edit file first.')
+            return
+        }
+
+        const clipboardContents = await vscode.env.clipboard.readText()
+
+        if (clipboardContents === '') {
+            return
+        }
+
+        if (clipboardContents.split('\n').length === 1) {
+            const fpath = path.resolve(fileUri.fsPath, clipboardContents)
+            if (fs.existsSync(fpath)) {
+                this.pasteFile(editor, fileUri.fsPath, clipboardContents)
+                return
+            }
+        }
+        // if not pasting file
+        try {
+            this.pasteTable(editor, clipboardContents)
+        } catch (error) {
+            this.pasteNormal(editor, this.reformatText(clipboardContents))
+        }
+    }
+
+    public pasteNormal(editor: vscode.TextEditor, content: string) {
+        editor.edit(edit => {
+            const current = editor.selection
+
+            if (current.isEmpty) {
+                edit.insert(current.start, content)
+            } else {
+                edit.replace(current, content)
+            }
+        })
+    }
+
+    public pasteFile(editor: vscode.TextEditor, baseFile: string, file: string) {
+        const TABLE_FORMATS = ['.csv']
+        const extension = path.extname(file)
+
+        if (TABLE_FORMATS.indexOf(extension) !== -1) {
+            if (extension === '.csv') {
+                const rows: string[] = []
+
+                fs.createReadStream(path.resolve(baseFile, file))
+                    .pipe(csv())
+                    .on('data', (data: Object) => rows.push(Object.values(data).join('\t')))
+                    .on('end', () => {
+                        const body = rows.join('\n')
+                        this.pasteTable(editor, body)
+                    })
+            }
+        }
+    }
+
+    public pasteTable(editor: vscode.TextEditor, content: string) {
+        this.extension.logger.addLogMessage('Pasting: Table')
+        // trim surrounding whitespace
+        content = content.replace(/^\s*/, '').replace(/\s*$/, '')
```

Comment:
Why not using `trim()`?

Reply:
The variable `content` seems to be a multiline string. Should the trimming be applied to each line?

Reply:
Hmm. Thinking about this a bit more we want to trim all white space other than tabs (because blank cells at the start/end of a line are possible). I think 0fdfc7cd1f512d90da202b49742943fe9f90b644 should do the trick.